corehq/apps/users/device_rate_limiter.py

```diff
@@ -0,0 +1,93 @@
+from datetime import datetime, timezone
+
+from django.conf import settings
+from django_redis import get_redis_connection
+
+from corehq import toggles
+from corehq.apps.cloudcare.const import DEVICE_ID as CLOUDCARE_DEVICE_ID
+from corehq.util.metrics import metrics_counter, metrics_histogram
+
+DEVICE_RATE_LIMIT_MESSAGE = "Current usage for this user is too high. Please try again in a minute."
+DEVICE_SET_CACHE_TIMEOUT = 2 * 60  # 2 minutes
+
+
+class DeviceRateLimiter:
+    """
+    Operates on a time window of 1 minute
+    """
+
+    def __init__(self):
+        # need to use raw redis connection to use srem and scard functions
+        self.client = get_redis_connection()
+
+    def device_limit_per_user(self, domain):
+        if toggles.INCREASE_DEVICE_LIMIT_PER_USER.enabled(domain):
+            return settings.INCREASED_DEVICE_LIMIT_PER_USER
+        return settings.DEVICE_LIMIT_PER_USER
+
+    def rate_limit_device(self, domain, user_id, device_id):
+        """
+        Returns boolean representing if this user_id + device_id combo is rate limited or not
+        NOTE: calling this method will result in the device_id being added to the list of used device_ids
+        """
+        if not device_id or self._is_formplayer(device_id):
+            # do not track formplayer activity
+            return False
+
+        key = self._get_redis_key(user_id)
+
+        if not self._exists(key):
+            self._track_usage(key, device_id, is_key_new=True)
+            return False
+
+        if self._device_has_been_used(key, device_id):
+            return False
+
+        device_count = self._device_count(key)
+        if device_count < self.device_limit_per_user(domain):
+            self._track_usage(key, device_id)
```

Comment:
I think this will hit redis at least 3 times and at most 5 times.

The number of hits could be reduced to 1 or 2, and the operation made atomic with a pipeline

```py
pipe = client.pipeline()
pipe.exists(key)
pipe.sadd(key, device_id)
pipe.scard(key)
pre_existing, added, count = pipe.execute()
if pre_existing:
    return added and count >= devices_allowed
client.expire(key, DEVICE_SET_CACHE_TIMEOUT)
return False
```

Reply:
Thanks for mentioning pipelines. Didn't know about this. As we discussed, this code as is would change behavior from what is desired currently. I adopted pipelines in c90a5bd33b991b78260d361c8ec1d1f94c23fbb2 to help minimize connections to redis and ended up with at most 2 calls to redis while keeping the current behavior.